File: src/crypto/ripemd160.ts
/**
 * RIPEMD160 Hash Function
 *
 * Legacy hash function used primarily in Bitcoin for address generation.
 * RIPEMD160 produces 20-byte (160-bit) hashes.
 *
 * @example
 * ```typescript
 * import { Ripemd160 } from './ripemd160.js';
 *
 * // Hash bytes
 * const hash = Ripemd160.hash(data);
 *
 * // Hash string
 * const hash2 = Ripemd160.hashString("hello");
 * ```
 */

import { ripemd160 as nobleRipemd160 } from "@noble/hashes/legacy.js";

// ============================================================================
// Ripemd160 Namespace
// ============================================================================

export namespace Ripemd160 {
	/**
	 * Compute RIPEMD160 hash (20 bytes)
	 *
	 * @param data - Input data (Uint8Array or string)
	 * @returns 20-byte hash
	 *
	 * @example
	 * ```typescript
	 * const hash = Ripemd160.hash(data);
	 * // Uint8Array(20)
	 * ```
	 */
	export function hash(data: Uint8Array | string): Uint8Array {
		if (typeof data === "string") {
			return hashString(data);
		}
		return nobleRipemd160(data);
	}

	/**
	 * Compute RIPEMD160 hash of UTF-8 string
	 *
	 * @param str - Input string
	 * @returns 20-byte hash
	 *
	 * @example
	 * ```typescript
	 * const hash = Ripemd160.hashString("hello");
	 * // Uint8Array(20)
	 * ```
	 */
	export function hashString(str: string): Uint8Array {
		const encoder = new TextEncoder();
		const bytes = encoder.encode(str);
		return nobleRipemd160(bytes);
	}
}

// Re-export namespace as default
export default Ripemd160;

File: packages/voltaire-go/crypto/ripemd160/ripemd160.go
// Package ripemd160 provides RIPEMD-160 hashing.
//
// RIPEMD-160 produces a 160-bit (20-byte) hash and is commonly used in
// Bitcoin for address derivation via hash160 (RIPEMD160(SHA256(data))).
package ripemd160

import (
	"crypto/sha256"
	"encoding/hex"
	"hash"

	"golang.org/x/crypto/ripemd160"
)

// Size is the size of a RIPEMD-160 hash in bytes.
const Size = 20

// BlockSize is the block size of RIPEMD-160 in bytes.
const BlockSize = 64

// Hash computes the RIPEMD-160 hash of data.
func Hash(data []byte) [Size]byte {
	h := ripemd160.New()
	h.Write(data)
	var result [Size]byte
	copy(result[:], h.Sum(nil))
	return result
}

// HashString computes the RIPEMD-160 hash of a UTF-8 string.
func HashString(s string) [Size]byte {
	return Hash([]byte(s))
}

// HashToHex computes the RIPEMD-160 hash and returns it as a hex string
// with 0x prefix.
func HashToHex(data []byte) string {
	h := Hash(data)
	return "0x" + hex.EncodeToString(h[:])
}

// Hash160 computes RIPEMD160(SHA256(data)), commonly used in Bitcoin
// for public key hashing and script hashing.
func Hash160(data []byte) [Size]byte {
	sha := sha256.Sum256(data)
	return Hash(sha[:])
}

// Sum is an alias for Hash, provided for API consistency.
func Sum(data []byte) [Size]byte {
	return Hash(data)
}

// New returns a new hash.Hash computing the RIPEMD-160 checksum.
func New() hash.Hash {
	return ripemd160.New()
}
